---
layout: tutorial_slides
logo: "GTN"

title: "Trajectory analysis"
questions:
  - "What is trajectory analysis?"
  - "What are the methods of trajectory inference and trajectories patterns?"
  - "How are the decisions about the trajectory made?"
  - "What to take into account when choosing the method for your data?" 
objectives:
  - "Becoming familiar with the methods of trajectory inference"
  - "Learning how the decisions are made by the algorithms"
  - "Being able to choose the method appropriate for specific data"
  - "Insight into methods available in Galaxy"
time_estimation:
subtopic: writing
key_points:
  - ""
contributors:
  - wee-snufkin
---

### What is trajectory analysis?
.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

Trajectory inference (TI) methods have emerged as a novel subfield within computational biology to better study the underlying dynamics of a biological process of interest, such as:

cellular development    | differentiation      | immune responses
:---:| :---:| :---:
.image-40[ ![development](../../images/scrna-casestudy-monocle/1_cell_delevopment.png) ] | .image-40[ ![differentiation](../../images/scrna-casestudy-monocle/2_differentiation.png) ] | .image-40[ ![immune](../../images/scrna-casestudy-monocle/3_immune.png) ]

--
TI allows studying how cells evolve from one cell state to another, and subsequently when and how cell fate decisions are made.

.image-40[ ![process](../../images/scrna-casestudy-monocle/4_process.png) ]
---

### Clustering, trajectory and pseudotime
.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

.pull-left[
**Clustering** calculates cell similarities to identify group cell types, that can be identified based on the marker genes expressed in each cluster. 

.image-75[ ![clusters](../../images/scrna-casestudy-monocle/5_clusters.png) ]

]
--
.pull-right[
**Trajectory inference** helps to understand how those cell types are related - whether cells differentiate, change in response to stimuli or over time. To infer trajectories, we need data from cells at different points along a path of differentiation. This inferred temporal dimension is known as pseudotime. 
**Pseudotime** measures the cells’ progress through the transition.

.image-60[ ![trajectory](../../images/scrna-casestudy-monocle/6_trajectories.png) ]

]

---
### Trajectory inference approaches: **Clustering-based (tree based)**

.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

- Try to find stable cell states in the data and afterward connect these states to form a trajectory and map the cells to the formed graph structure 

--

- Clustering approaches: soft K-means, Louvain, non-negative matrix factorization, hierarchical clustering

--

- To connect the clusters to each other, usually a **minimum weight spanning tree (MST)** is constructed (connecting without any cycles and with the minimum possible total edge weight)

--

.image-25[ ![mst](../../images/scrna-casestudy-monocle/mst.png) ]
---
### Trajectory inference approaches: **Graph-based**
.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

- Construct a graph representation of the cells and either use graph decomposition to reveal connected and disconnected components, or graph diffusion or traversal methods to construct the trajectory topology

--

- Most methods start by building a **k-nearest neighbors (kNN) ** graph using the Euclidean distance

.image-50[ ![knn](../../images/scrna-casestudy-monocle/knn.png) ]
--

- Diffusion peudotime (DPT) calculates the probability of cells transitioning into each other using random walks from a user-provided root cell and uses the differences between these probabilities as pseudotime


---

### Trajectory inference approaches: **Manifold learning-based**
.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

- Based on the assumption that data lie on a low-dimentional manifold embedded in a higher-dimentional space

--

- Using principal curves and graphs to obtain smooth trajectories

--

- Usually used dimentionality reduction methods: **t-SNE** (t-distributed stochastic neighbor embedding), **UMAP** (Uniform Manifold Approximation and Projection), **PCA** (Principal component analysis), **ICA** (Independent Component Analysis), **LSA** (Latent semantic analysis)

.image-60[ ![dim red](../../images/scrna-casestudy-monocle/dim_red_methods.png) ]

---

### Extensions of trajectory inference: **RNA velocity methods**
.footnote[[Deconinck et al. Recent advances in trajectory inference from single-cell omics data](https://doi.org/10.1016/j.coisb.2021.05.005.)]
--

- RNA velocity methods estimate the future state of a single cell captured in a static snapshot by looking at the ratios between spliced mRNA, unspliced mRNA, and mRNA degradation. 

--

- scVelo and CellRank use this extra velocity information to construct a directed k-nearest neighbor graph, as a starting step for the TI method. 

--

- This has the advantage that root cell specification is not necessary, and adds directional information to the trajectory.

--

.image-25[ ![velocity](../../images/scrna-casestudy-monocle/velocity.png) ]

---
### Assumptions
--

- It would be quite difficult to analyse the sample every several seconds to see how the cells are changing. Therefore, we assume that snapshot data encompass all naïve, intermediate, and mature cell states with sufficient sampling coverage to allow the reconstruction of differentiation trajectories.    
[Sagar and Grün, 2020](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7115822/)

--

- As different TI methods make different assumptions about the data, a first choice to make is based on which biological process is to be expected: **not all TI methods are designed to infer all kinds of biological processes.**  
[Deconinck et al., 2021](https://doi.org/10.1016/j.coisb.2021.05.005)

---
### Why not all TI methods are designed to infer all kinds of biological processes?
--

Look at the pipeline presented by [Robrecht Cannoodt et al., 2016](https://onlinelibrary.wiley.com/doi/10.1002/eji.201646347)

--

.image-100[ ![pipeline](../../images/scrna-casestudy-monocle/7_methods.jpg) ]        
--

There are multiple methods using particular algorithms, or even their combinations, so you must consider which one would be best for analysing your sample.

---

### When analysing your data, consider the following: 
--

.pull-left[  
  

- Tissue from which the cells come from   

- Branching points  

- Supervised and unsupervised learning  

- Format of the data  

- Number of cells and features  

- Computing power & running time   

]

--

.pull-right[
    
 
To help you evaluate which method would work best for your data, check out this [**awesome comparision site - dynguidelines**](https://zouter.shinyapps.io/server/), a part of a larger set of open packages for doing and interpreting trajectories called [the dynverse](https://github.com/dynverse/dynverse).     
  
.image-100[ ![zouter](../../images/scrna-casestudy-monocle/zouter.jpg) ]

]

---

### When analysing your data, consider the following: **Tissue from which the cells come from**
--

.pull-left[
    
  
  
- Gives an idea about the topology you expect and the priors you can feed the method with, as well as expected cell types and potential contaminants  

- You need to know about 75% of your data and make sure your analysis shows that, for you to then identify the 25% new information  

- Trajectory analysis is quite a sensitive method, so always check if the obtained computational results make biological sense!

]

.pull-right[  
  

![scheme](../../images/scrna-casestudy-monocle/8_scheme.png)  
<sub><sub><sub> Reprinted from “Heterogeneity of Murine Lymph Node Stromal Cell Subsets”, by BioRender.com (2022) </sub></sub></sub>
]

---
### When analysing your data, consider the following: **Branching points**

--

Cells can differentiate or develop in various way, so they may exhibit different topologies. 

--

![table](../../images/scrna-casestudy-monocle/table.jpg)  

---
### When analysing your data, consider the following: **Branching points**

explain more, add binning

---

### When analysing your data, consider the following: **Supervised and unsupervised learning**
--

.pull-left[  
  

![root_cells](../../images/scrna-casestudy-monocle/root_cells.jpg)  

]

--

.pull-right[
    
 
- Some algorithms work fully unsupervised, ie. the user is not required to input any priors.  

- However, many of them take the ‘starting cell’, or ‘end cell’ as information that helps to infer the trajectory which best represents the actual biological processes.  

- On the other hand, priors can bias the outcome of the method, but if chosen appropriately, they are really helpful.  

]

---
### When analysing your data, consider the following: **Supervised and unsupervised learning**

If you know which cells are root cells, you should enter this inforamtion to the method to make the computations more precise. However, some methods use unsupervised algorithms, so you will get a trajectory based on the tools they use and topology they can infer.    

--

Unsupervised | Priors needed: start cells | Priors needed: end cells | Priors needed: both start and end cells 
 :---:| :---:| :---: | :---: 
Slingshot, SCORPIUS, Angle, MST, Waterfall, TSCAN, SLICE, pCreode, SCUBA, RaceID/StemID, Monocle DDRTree | PAGA Tree, PAGA, Wanderlust, Wishbone, topslam, URD, CellRouter, SLICER | MFA, GrandPrix, GPfates, MERLoT | Monocle ICA

---
### When analysing your data, consider the following: **Format of the data**
--

- You have to check that your chosen method is compatible with the format of your data. If not, consider converting it or even contacting the developers to implement this into the pipeline.  

--

- Other possible inputs: 
 - Raw (FASTQ); 
 - MMformat (output of 10x); 
 - EBI-SCXA; 
 - AnnData   
   
--

- Powerful [SCEasy tool](https://toolshed.g2.bx.psu.edu/repository/display_tool?repository_id=a4381cd7162476d9&render_repository_actions_for=tool_shed&tool_config=%2Fsrv%2Ftoolshed%2Fmain%2Fvar%2Fdata%2Frepos%2F004%2Frepo_4830%2Fsceasy_convert.xml&changeset_revision=f62bc418173f) in Galaxy to convert between formats

- Implementations  

![implementation](../../images/scrna-casestudy-monocle/implementation.jpg)  

---


Slides begin with a title `### ...` and should end with a `---`

???

- Slides are written with standard markdown, and begin with a header, and end with three dashes.
- Lists, links, and images are all written with their normal syntax.
- If you're familiar with markdown and already writing your tutorial in it, this should be an easy addition.
- The only differences are the dashes to separate slides, and speaker notes which we'll cover later.

---



### Formatting: Alignment

.left[Left-aligned text]
.center[Centered text]
.right[Right-aligned text]
.left[![Left-aligned image]({{ site.baseurl }}/assets/images/GTN-60px.png)]
.center[![Centered image]({{ site.baseurl }}/assets/images/GTN-60px.png)]
.right[![Right-aligned image]({{ site.baseurl }}/assets/images/GTN-60px.png)]

```markdown
.left[Left-aligned text]
.center[Centered text]
.right[Right-aligned text]
.left[![Left-aligned image]({{ site.baseurl }}/assets/images/GTN-60px.png)]
.center[![Centered image]({{ site.baseurl }}/assets/images/GTN-60px.png)]
.right[![Right-aligned image]({{ site.baseurl }}/assets/images/GTN-60px.png)]
```

???

- Everything can be aligned using several CSS classes which are provided, left, right, and center.
- The classes are added with a dot followed by the class, and then the contents within square brackets.

---



### Formatting: Incremental Text

1. Item

???

- Incremental text can be added

--

2. Appears

???

- And appears incrementally on every slide.

--


Just separate every item with two dashes (`--`)

```markdown
1. Item

--

2. Appears
```

???

- You can accomplish this with two dashes instead of three for incremental slides.
- The dashes should be on their own line, just like the slide separating dashes.

---



### Formatting: Images

.pull-left[
Demo:

10%

.image-10[ ![GTN Logo (tiny)]({{ site.baseurl }}/assets/images/GTN-60px.png) ]

50%

.image-50[ ![GTN Logo (slightly larger)]({{ site.baseurl }}/assets/images/GTN-60px.png) ]

]
.pull-right[
Source:

```markdown
10%
.image-10[
  ![GTN Logo]({{ site.baseurl }}/assets/images/GTN-60px.png)
]

50%
.image-50[
  ![GTN Logo]({{ site.baseurl }}/assets/images/GTN-60px.png)
]
```
]


There are [other CSS classes available](https://github.com/galaxyproject{{ site.baseurl }}/blob/main/assets/css/slides.css#L67) for other sizes

???

- There are a number of CSS classes available to use for styling the content you show on your slides.
- Here you can see images scaled to 10% and 50%.
- There are a number of other classes that you can find in the GTN's codebase.


---



### Formatting: Tables

.pull-left[
Demo:

column1    | column2      | column3
---        | ---          | ---
aaaaaaaaaa | bbbb         | cccc
dddd       | eeeeeeeeeeee | ffff
gggg       | hhhh         | iiiiiiiiii
]

.pull-right[
Source:

```markdown
column1    | column2      | column3
---        | ---          | ---
aaaaaaaaaa | bbbb         | cccc
dddd       | eeeeeeeeeeee | ffff
gggg       | hhhh         | iiiiiiiiii
```
]

By default, tables are left-aligned with text left-aligned.

???

- Tables can be written in the simplified markdown syntax, or using a more complex one that allows for alignment.
- Here a simple table is shown with several columns.
- The tables do not need to be aligned, we just did that in this slide for clarity.

---



### Formatting: Tables with Aligned Text

Different alignments for the individual columns can be controlled with the `:` character

.pull-left[
Demo:

| Tables       | Are      | Cool          |
|:-------------|:--------:|--------------:|
| left-aligned | centered | right-aligned |
| row2         | row2     | row2          |

]

.pull-right[
Source:
```markdown
| Tables       | Are      | Cool          |
|:-------------|:--------:|--------------:|
| left-aligned | centered | right-aligned |
| row2         | row2     | row2          |
```
]

???

- If you need more complex tables, you can use the colon character to control the alignment.
- If you are presenting numeric data or similar then this can be useful to make the data more readable.

---



### Formatting: Footnotes

Content with footnotes

Source:
```markdown
.footnote[Normal footnote with plenty of text]
```

Demo:

.footnote[Normal footnote with plenty of text]

???

- Footnotes can be used for explanatory notes or adding things like references.
- References are ideal footnote content as they are not necessarily important for comprehension of the slides, but might be useful later for learners.

---



### Formatting: Two Columns

This is achived with the `.pull-left` and `.pull-right` classes.

```markdown
.pull-left[
Some text, as many texts as we want...

or even table

column1 | column2 | column3
--- | --- | ---
row1 | r1c2 | r1c3
row2 | r2c2 | r2c3
row3 | r3c2 | r3c3
]

.pull-right[ ![GTN Logo]({{ site.baseurl }}/assets/images/GTN-60px.png) ]
```

Will be rendered as:

???

- Two column layouts can easily be achieved with the pull left and pull right classes.
- The result of this is demonstrated on the next slide.


---



### Formatting: Two Columns

.pull-left[
Some text, as many text as we want.....

or even table

column1 | column2 | column3
--- | --- | ---
row1 | r1c2 | r1c3
row2 | r2c2 | r2c3
row3 | r3c2 | r3c3
]

.pull-right[
![GTN Logo]({{ site.baseurl }}/assets/images/GTN-60px.png)
]

???

- Any content can be placed in the left and right columns
- The addition of these classes will wrap all of their contents and place them on the left or right sides.
- The two column layout is an especially good choice if you want to show a graphic and discuss some important points related to it.

---



## Presenter Notes

Allows you to:

- See the next slide before the audience
- See notes you have made on your slides

![Image showing presenter mode](../../images/presenter-mode.png)

???

- Presenter Mode lets you have two windows with the same slides, one with presenter notes and a preview of the upcoming slide.
- When giving presentations with multiple screens or on a projector, this makes life less stressful.
- You can use the presenter notes to remind you, and others, of important content to mention on each slide.

---



### Presenter Notes: Formatting

.left[
<br/>
&#35;&#35;&#35; My Slide
<br/><br/>
Some text!
<br/><br/>
&#63;&#63;&#63;
<br/><br/>
Things written below the ??? are only shown in the
presenter view. Press `p` to bring this up
<br/><br/>
&#45;&#45;&#45;
<br/><br/>
]

???

- Things written below the three question marks are only shown in the presenter view.
- Press P to bring this up, and the question mark to show the presenter view shortcut keys.


---



## Shortcut Keys

- <kbd>?</kbd> activates help
- <kbd>Escape</kbd> leaves help
- <kbd>p</kbd> activates presenter mode
- <kbd>f</kbd> activates full screen mode
- <kbd>c</kbd> clones the slides into a popup

???

- The key to close the help isn't shown in the keyboard shortcuts, but it is the escape key.
- There are a number of other shortcuts which can be useful to know about.
- The clone window function creates a popup with a copy of the slides, this can be used to show the presenter notes on one screen, and the actual slides on another screen that learners can see.

---

## Videos

.pull-left[
- Slides with good speaker notes can be turned into videos
- Enable this in the metadata of the slides:

  ```markdown
  title: "Creating Slides"
  video: true
  questions:
    - "How to format slides?"
  ```

- automatically rendered into narrated videos by the GTN
]

.pull-right[
.image-100[![Screenshot of GTN video player showing a set of slides with a play icon and a transcript](../../images/slides-to-video.png)]
]

???

- The GTN automatically creates videos from selected slide decks.
- Simply by setting video to true to the metadata, slides with sufficiently good speaker notes can be automatically turned into videos.
- The slides have audio narration spoken from the script of the speaker notes by AWS Polly.
- The audio clips, and slides are combined into a video which is then available in the GTN Video Library
