---
layout: tutorial_slides
logo: "GTN"

title: "Identification of non-canonical ORFs and their potential biological function"
contributors:
  - gallardoalba

priority: 4

video: no
zenodo_link: ""


---

## Index of contents

1. Introduction

2. Galaxy workflow

---

## Introduction

- What do we mean by non-canonical ORFs?

- Why study non-canonical ORFs?

- Why have not been yet characterized?

- How identify non-canonical ORFs?

---

## What do we mean by non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/cambridge_dictionary.png %})

.footnote[Source: [Cambridge dictionary](https://dictionary.cambridge.org/dictionary/english/canonical)]

---

## What do we mean by non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/expression_distribution.png %})

.footnote[Source: Erady, Chaitanya, et al. "Pan-cancer analysis of transcripts encoding novel open-reading frames (nORFs) and their potential biological functions." NPJ Genomic Medicine 6.1 (2021): 4.]

---

## What do we mean by non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/dark_proteome.png %})

---

## Why study non-canonical ORFs?

- Potentially novel <b>prognostic and diagnostic markers</b>

- The vast majority <b>have not been investigated</b>

- Particulary attractive as <b>allosteric celullar regulators</b>

---

## Why study non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/non-canonical_nature.png %})

---

## Why have not been characterized?


- Arbitrary thresholds on ORF lengths
    - Peptides smaller than 100 aminoacids are usually discarted
- Frequently annotated as non-coding RNAs
- Propensity for structural disorder
    -  Discarted as intrinsically disordered proteins (IDPs)

---

## Why study small peptides?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/small_peptides.png %})

---

## Why study small peptides?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/short_peptides.png %})

---

## Why study small peptides?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/small_peptides_examples.png %})

.footnote[Source: Steinberg, R., & Koch, H. G. (2021). The largely unexplored biology of small proteins in pro- and eukaryotes. The FEBS journal, 288(24), 7002-7024]

---

## Annotated as non-coding RNAs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/question.jpg %})

---

## Annotated as non-coding RNAs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/small_non_coding.png %})

---

## Annotated as non-coding RNAs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/spencer_datase.png %})

---

## Why study intrinsically disordered proteins (IDP)?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/disordered_regions.png %})

.footnote[Source: Babu, M. M., Kriwacki, R. W., & Pappu, R. V. (2012). Versatility from protein disorder. Science, 337(6101), 1460-1461.]

---

## Disorder-Function Paradigm

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/disorder_article.png %})

---

## Disorder-Function Paradigm

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/intrinsically_disordered.png %})

---

## Disorder-Function Paradigm

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/disorder_protein.jpg %})

.footnote[Source: Chakrabarti, P., & Chakravarty, D. (2022). Intrinsically disordered proteins/regions and insight into their biomolecular interactions. Biophysical Chemistry, 283, 106769]

---

## How identify non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/scientist.jpg %})

---

## How identify non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/isoform_switching.png %})

---

## How identify non-canonical ORFs?

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/isoform_switching.jpg %})

---

## Galaxy Workflow

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/galaxy_training.png %})

Full detailed explanation in the [Genome-wide alternative splicing analysis](https://gxy.io/GTN:T00345) Galaxy training.

---

## Galaxy Workflow

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/full_workflow.png %})

Full detailed explanation in the [Genome-wide alternative splicing analysis](https://gxy.io/GTN:T00345) Galaxy training.

---

## Galaxy Workflow

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/simple_workflow.png %})

---

## Initial QC assessment

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/initial_QC.png %})

Identify potential artifacts that may impact the interpretation of downstream analysis.

---

## Mapping and identication of novel splicing sites with RNASTAR

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/STAR_pipeline.png  %})

Two-pass alignment enables sequence reads to span novel splice junctions by fewer nucleotides, conferring greater read depth and providing significantly more accurate quantification of novel splice junctions.

---

## Post-mapping QC assessment with RSeQC

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/rseqc_junction_saturation_plot.png %})

RSeQC is a toolkit for generating RNA-seq-specific quality control metrics. The figure corresponds to RSeQC junction saturation of known (A) and novel (B) splicing sites.

---

## Reference-based transcriptome assembly and quantification with StringTie

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/stringtie_pipeline.png %})

StringTie is a fast and highly efficient assembler of RNA-seq alignments into potential transcripts.

---

## Post-assembly QC assessment with rnaQUAST

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/rnaQUAST_cumulative_isoform.png %})

rnaQUAST, which will provide us diverse completeness/correctness statistics very useful in order to identify and address potential errors or gaps in the assembly process. The figure is a rnaQUAST cummulative isoform plot.

---

## Isoform switching and functional analysis with IsoformSwitchAnalyzeR

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/isoformswitch_pipeline.png %})

IsoformSwitchAnalyzieR performs the differential isoform usage analysis by using DEXSeq.

---

## Isoform switching and functional analysis with IsoformSwitchAnalyzeR

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/consequences_plot.png %})

To analyze large-scale patterns in predicted IS consequences, IsoformSwitchAnalyzeR computes all isoform switching events resulting in a gain/loss of a specific consequence (e.g. protein domain gain/loss)

---

## Candidate non-canonical ORF

![example_text]({% link topics/transcriptomics/images/differential_isoform/slides/candidate_gene.png %})

