---
layout: event
title: "Mycobacterium tuberculosis complex NGS made easy"

description: |
  These series of webinars and tutorials aim at improving basic and
  applied knowledge associated with next-generation sequencing (NGS)
  technologies and their applications in the field of Tuberculosis (TB).

cover-image: events/images/TBCAPT_NGS_Training_banner.jpg
cover-image-alt: banner for the course

date_start: 2023-10-02
date_end: 2023-10-06 # optional, if event is more than one day

contributions:
  organisers: # GTN contributors or funders, must be defined in CONTRIBUTORS.yaml
  - dbrites
  instructors:
  - dbrites
  - cstritt

contact_email: d.brites@swisstph.ch

location:
  name: Online

cost: free
audience:  Open for all, but target audience is clinicians and researchers using MTBC sequencing data.

async: false
mode: online

program:
  - section: "Day 1: Overview of NGS technologies & TB specific NGS solutions"
    description: |
      There were unusually high rates of TB cases in your country this year. To characterize the underlying
      bacterial strains driving the epidemic, isolates have been sent for whole-genome sequencing.  Doctors and public
      health authorities request information in order to take decisions. In this course it will be demonstrated how
      you would make use of NGS to answer several questions relevant for patient and public health system management
      such as:
        - Are there cases of drug resistant bacteria?
        - Is there transmission of drug resistance?
        - Is there evidence of de novo emergence of resistance?
        - Are there multiple infections per patient?
        - Do we have on-going transmission?

      We hope that at the end of the different training sessions you can answer this question on your own and can apply what you have learnt to your own data!

  - tutorials:
    - type: custom
      name: Icebreaker
      description: |
        Introduce yourself on Slack and tell us one fun fact about yourself.
        Post your answer to the [course Slack channel](https://gtnsmrgsbord.slack.com/archives/event-mtb-ngs)
        Please feel free to respond to each other here, this channel is for socializing and getting to know each other! :)

    - type: custom
      name: "Webinar: Overview of NGS technologies & TB specific NGS solutions"
      description: |
        This [webinar](https://www.youtube.com/watch?v=BcfBjCJRmUs) will introduce different sequencing technologies
        and what applies best to what kind of problem.

        Finished watching the webinar as part of the Mtb NGS training event? Share your thoughts with us
        [here](https://docs.google.com/document/d/1EfgttagZ8hBfjQiRznsCgDYVpzi3CF55cHqjarYa97o/edit#heading=h.e69aqv1nq3gw)

    - type: custom
      name: "Webinar: Implementation of NGS for TB- WHO documents and other considerations"
      description: |
        [This webinar](https://youtu.be/eSWnPange8w) will summarize the recommendations and considerations
        available from the WHO documents on the use of NGS for TB

    - type: custom
      name: Q&A session
      time: "14:00-16:00"
      description: |
        If you are part of the Mtb NGS training event we would like to hear your opinion on the following questions
        and promote discussion in this Q&A session. Please let us know your thoughts on one or more of the following
        questions in the shared notes;

        1. How do I choose the right sequencing technology for my samples?
        2. What is needed for NGS? 3) Why is NGS better for drug resistance and outbreak analysis?

        Also take the chance to ask us or write down in the shared notes other questions you might have.

        - Zoom: [link](https://swisstph.zoom.us/j/88614333103?pwd=OVVjeG04a2lVMXB0THUxMXoxNUNkdz09)
        - Shared Notes: [link](https://docs.google.com/document/d/1VtGho4Zz0qFp5tzqD8rl7NU83xtYb4tCkLDitC3DrhQ/edit#)

        *Note: The completion of this assessment is a requirement for the certificate of attendance.*

  - section: "Day 2"
    description: |
      The 20 strains isolated in your country have been sequenced with
      Illumina technology to obtain whole-genome sequences. In this part
      of the workshop you will learn how to analyse those sequences.

      In a typical bioinformatic pipeline you would store your sequences
      in a computer server where all necessary software would be
      installed. This would be a server running the operating system
      LINUX, which is the most efficient way to run bioinformatics
      pipelines (more on this on Day 4). You will be running your
      analysis in a LINUX server from Galaxy, but instead of writing
      directly commands to execute operations in the server, you will be
      executing operations through a Galaxy graphical interface. This
      allows you to have access to a LINUX server and to run workflows
      without knowing LINUX. Importantly, for training purposes it also
      allows you to dedicate more attention in trying to understand what
      is being done in each of the steps without having to understand the
      programing behind. This being said, working directly on a LINUX
      cluster provides you always more flexibility, but if you donâ€™t have
      access to one, Galaxy is a very good alternative for data analysis.

      You will need to understand how to use Galaxy to run all the hands-
      on tutorials and therefore is highly recommended that you follow
      the next webinar and hands-on on Galaxy. The good thing about this
      is that once you know how it works, you can use it to run your own
      analysis with your own data.

    tutorials:
      - name: Icebreaker
        type: custom
        description: |
           Come say Hi in Slack! Let us know you are joining today and are getting started!
           Today's icebreaker question:
           - What is your favorite dish (food or drink)? Bonus points for recipes!

           Post your answers in #event-mtb-ngs on Slack!

           Please feel free to respond to each other here, this channel is for socializing and getting to know each other! :)






---

![banner for course]({% link events/images/TBCAPT_NGS_Training_banner.jpg %})

These series of webinars and tutorials aim at improving basic and applied knowledge associated with next-generation sequencing (NGS) technologies and their applications in the field of Tuberculosis (TB).

The trainings will introduce scalable and reproducible data analysis with Galaxy of *Mycobacterium tuberculosis* complex (MTBC) genomes.

A series of pre-recorded sessions and hands-on tutorials will show:
  - How to differentiate sequencing technologies and which ones are most commonly applied in TB and how **(Day 1)**
  - How NGS can be implement into TB laboratories **(Day 1)**
  - How to do mapping and variant calling **(Day 2)**
  - How to detect drug resistance conferring mutations, build phylogenetic trees and infer tuberculosis transmission **(Day 3)**
  - How to use different web tools dedicated to targeted analysis and what it takes to do bioinformatics **(Day 4)**.

After the trainings all participants are expected to:
  - Describe how NGS is being used in TB research and clinical pratice
  - Extract MTBC genomic variants from short sequencing reads
  - Identify drug resistant mutations
  - Identify genetic relationships and interpret a phylogenetic trees
  - Use web applications designed for M. tuberculosis
